In [1]:
from IPython.display import HTML # Versteckt Code auf Knopfdruck

HTML('''<script>
    function code_toggle() {
        if (code_shown){
            $('div.input').hide('500');
            $('#toggleButton').val('Show Code')
        } else {
            $('div.input').show('500');
            $('#toggleButton').val('Hide Code')
        }
        code_shown = !code_shown
    }

    $(document).ready(function(){
        code_shown = false;
        $('div.input').hide()
    });
    </script>
    <form action = "javascript:code_toggle()">
    <input type = "submit" id = "toggleButton" value = "Show Code">
    </form>''')

# Epidemie


- **Datum:** 06.05.2024
- **Autor:** Natalia Kotkova, Dino Dervisevic

## Kritik aus dem ersten Projekt (Dino)
- Der Code ist sehr umfangreich, evtl. In Module zerlegen und diese besser beschreiben
- Es würde helfen, eine Zusammenfassung aller Ergebnisse in einer Tabelle (Anfang oder Ende) anzubieten / UnitTests
- Der Code läuft fehlerfrei aber extrem langsam (Sage 10.2)
- Die Codeblöcke sind extrem lang und teilweise schwer verständlich
- Kaum Kommentare, um die lange Codeblöcke verständlicher zu machen
- Sehr langer Code ohne Herleitung

## Kritik aus dem ersten Projekt (Natalia)
- Die Problemdefinition sollte klarer herausgearbeitet werden 
- Es würde helfen, eine Zusammenfassung aller Ergebnisse in einer Tabelle (Anfang oder Ende) anzubieten 
- Es wurde keine Erklärung zum Vorgehen geliefert  4) Die Lösung ist recht kurz
- Keine Kommentare, um die lange Codeblöcke in Aufgabe 3 verständlicher zu machen
- Das Design des Tools ist kompliziert

# Aufgabe
Wir schreiben das Jahr 2115. In der Marskolonie New Atlantis ist das berüchtigte Beteigeuze Fieber ausgebrochen. New Atlantis wurde im Jahr 2055 gegründet und umfasst mittlerweile 50.000 Einwohner. Das Beteigeuze-Fieber ist eine lebensgefährliche, höchst ansteckende Krankheit, die von den galaktischen Läufern im Jahr 2099 in das Sonnensystem eingeschleppt wurde. Die galaktischen Läufer stammen vom 13. Planeten der Sonne Beteigeuze aus dem Sternbild Orion.

Das Beteigeuze-Fieber wird, ähnlich wie einige längst ausgestorbene Krankheiten der Erde, in 5 Stadien unterteilt. Die Stadien lauten:
- gesund
- infiziert
- krank
- schwer krank
- tot

Der Krankheitsverlauf des Fiebers läuft erfahrungsgemäß folgendermaßen ab:

- 10% der Gesunden werden pro Woche infiziert.
- 3% der Infizierten werden pro Woche gesund, 80% werden pro Woche krank und 6% schwer krank.
- 10% der Kranken werden schwer krank, 50% der Kranken werden gesund.
- 43% der Schwerkranken wechseln im Zuge der Genesung zurück in das Stadium krank, 9% der Schwerkranken sterben.

Wer die Krankheit einmal überlebt hat, ist immun und erkrankt nicht noch einmal. 3 % der Bevölkerung ist deswegen auf Grund der 1. Großen Beteigeuze-Epidemie im Jahre 2108 immun. Wir gehen davon aus, dass die New Atlantiker bevölkerungstechnisch stabil sind. Aus einer früheren Epidemie sind noch 1500 Menschen immun.

## Aufgabenstellung

### Aufgabe 1 – Modellieren Sie den Krankheitsverlauf
1. Schreiben Sie das Übergangsgesetz in Matrixform
2. Wie lautet die Krankheitsverteilung vier Wochen nach dem Ausbruch?
3. Wie viele Einwohner sind nach einem Jahr gestorben?
4. Zeichnen für die ersten 10 Wochen jeweils ein Diagramm?
5. Erstellen Sie eine Animation für die Bevölkerungsverteilung auf die Stadien für das erste Jahr (Sie können auch einen Slider für die Monate verwenden)

### Aufgabe 2 – Bestimmen Sie das Ender der Epidemie
1. Nach welcher Zeit ist die Epidemie überstanden? d.h. weniger als 0,1% der Bevölkerung ist krank.
2. Wie viele New Atlantiker werden insgesamt sterben? (Vorsicht diese Aufgabe ist unabhängig von der letzten.
3. Welche Kosten entstehen für die Epidemie, wenn ein kranker Atlantiker pro Woche 20 Space Coins kostet und ein schwer kranker Atlantiker 150?

### Aufgabe 3 – Einfluss von Impfung
Stellen Sie Überlegungen zum Einsatz einer Impfung an und bauen diese Überlegungen in Ihre Modelle ein. Eine Impfung kostet Geld (z.B. 15 Space Coins) und verringert die Wahr- scheinlichkeit zu erkranken oder schwer zu erkranken. (z.B. Wahrscheinlichkeit schwer zu erkranken wird um 50% reduziert). Diskutieren Sie, welchen Kriterien eine Impfung genügen müsste.



## Problemdefinition und gesuchte Werte
### Aufgabe 1 – Modellieren Sie den Krankheitsverlauf
1. **Übergangsgesetz in Matrixform**: Wir suchen eine Methode, um die Übergangswahrscheinlichkeiten zwischen den Krankheitsstadien (gesund, infiziert, krank, schwer krank, tot, immun) in einer Matrix darzustellen.
2. **Krankheitsverteilung nach vier Wochen**: Wir wollen die Verteilung der Bevölkerung auf die Krankheitsstadien vier Wochen nach Ausbruch der Epidemie ermitteln.
3. **Todesfälle nach einem Jahr**: Wir möchten berechnen, wie viele Menschen nach einem Jahr an der Krankheit gestorben sind.
4. **Diagramme für die ersten 10 Wochen**: Wir erstellen Diagramme, die die Verteilung der Bevölkerung auf die Krankheitsstadien in den ersten 10 Wochen zeigen.
5. **Bevölkerungsverteilung animieren**: Wir erstellen eine Animation, die die Veränderung der Krankheitsstadien der Bevölkerung im ersten Jahr zeigt, optional mit einem Slider für die Monate.

### Aufgabe 2 – Bestimmen Sie das Ende der Epidemie
1. **Ende der Epidemie**: Wir suchen den Zeitpunkt(die genaue Woche), an dem die Epidemie vorbei ist, d.h. wenn weniger als 0,1% der Bevölkerung krank sind.
2. **Gesamtzahl der Todesfälle**: Wir wollen ermitteln, wie viele Einwohner von New Atlantis insgesamt an der Epidemie sterben werden(also nache dem "biologischem Ende" der Epidemie).
3. **Kosten der Epidemie**: Wir berechnen die Kosten der Epidemie, wobei ein kranker Einwohner pro Woche 20 Space Coins und ein schwer kranker Einwohner 150 Space Coins kostet.

### Aufgabe 3 – Einfluss von Immunisierung
1. **Einsatz einer Impfung**: Wir überlegen, wie der Einsatz einer Impfung die Epidemie beeinflussen könnte und integrieren dies in unser Modell.
    - Wir berücksichtigen die Kosten der Impfung (z.B. 15 Space Coins pro Impfung).
    - Wir diskutieren, wie die Impfung die Wahrscheinlichkeit zu erkranken oder schwer zu erkranken reduzieren könnte (z.B. um 50%).
    - Wir überlegen, welche Kriterien eine Impfung erfüllen müsste, um effektiv zu sein und eine Kosten-Nutzen-Analyse zu rechtfertigen.

## Vorgaben
- 50000 Einwohner wovon 1500 immun sind
- 5 Stadien:
    - gesund
    - infiziert
    - krank
    - schwer krank
    - tot
- Krankheitsverlauf:
    - 10% der Gesunden werden pro Woche infiziert.
    - 3% der Infizierten werden pro Woche gesund, 80% werden pro Woche krank und 6% schwer krank.
    - 10% der Kranken werden schwer krank, 50% der Kranken werden gesund.
    - 43% der Schwerkranken wechseln im Zuge der Genesung zurück in das Stadium krank, 9% der Schwerkranken sterben.


## Allgemeine Annahmen[9]
- **Immunität**: Aus einer vorherigen Epidemie sind 1.500 Menschen immun und können sich nicht mehr anstecken oder krank werden.
    - Der Immun-Status wird als sechstes Stadium betrachtet und erhält einen eigenen Eintrag in der Markov-Matrix.
    - Die Bevölkerung besteht zu Beginn (Woche 0) aus 48.500 gesunden und 1.500 immunen Individuen.
- **Patient 0**: In der ersten Woche gibt es einen Infizierten.
- **Immunisierung durch Genesung**: Wenn kranke Menschen wieder gesund werden, werden sie auch immun. Dies wird im Eintrag (6,3) der Markov-Matrix berücksichtigt.
- **Markov-Matrix Annahmen**: Um die Eigenschaften der Markov-Matrix zu erfüllen (die Summe jeder Spalte muss eins ergeben), wurden Annahmen für die Wahrscheinlichkeiten getroffen.
    - Beispiel: Wenn 10% der gesunden Menschen infiziert werden, bleiben die restlichen 90% gesund.



# Lösungsansätze
## Allgemein
- Wir erstellen eine Klasse, die den Verlauf der Epidemie simuliert.

## Aufgabe 1 – Modellieren Sie den Krankheitsverlauf

### 1.1 Übergangsgesetz in Matrixform[1][5]
- Wir tragen die Übergangswahrscheinlichkeiten zwischen den Krankheitsstadien in die Markov-Matrix ein. Dabei stellen wir sicher, dass jede Spalte der Matrix zu 1 (bzw. 100%) summiert.

### 1.2 Krankheitsverteilung nach vier Wochen[3][5]
- Wir definieren einen Status Vektor, der den Zustand am Anfang der Epidemie repräsentiert.
- Wir definieren eine Hilfsfunktionen für die Simulation der Epidemie.
- Wir multiplizieren die Markov-Matrix mit dem Statusvektor. Das Ergebnis nach vier Iterationen (vier Wochen) ist der Statusvektor nach dieser Zeit.
- Der Status Vektor repräsentiert die verschiedenen Stadien in der Bevölkerung zum ausgewählten Zeitpunkt.

### 1.3 Todesfälle nach einem Jahr[3][5]
- Wir simulieren den Verlauf der Epidemie für ein Jahr (52 Wochen), indem wir die Multiplikation 52 Mal durchführen.
- Die Anzahl der Tode wird durch den Eintrag "tote" im Status Vektor ausgerechnet .

### 1.4 Diagramm für den Epidemieverlauf in den ersten zehn Wochen[4][7]
- Wir visualisieren den Verlauf der ersten 10 Wochen mit den Hilfsfunktionen unserer Klasse und der Matplotlib-Bibliothek.

### 1.5 Animation für die Bevölkerungsverteilung auf die Stadien für das erste Jahr[6][7]
- Wir animieren die Bevölkerungsverteilung auf die verschiedenen Stadien für das erste Jahr mit den Hilfsfunktionen unserer Klasse und der SageMath-Umgebung.


## Was verwenden wir, um das Problem zu lösen? 
### Aufgabe 1.1:  Markov Matrix[1][5]

$
P = \begin{pmatrix}
p_{11} & p_{12} & \cdots & p_{1n} \\
p_{21} & p_{22} & \cdots & p_{2n} \\
\vdots & \vdots & \ddots & \vdots \\
p_{n1} & p_{n2} & \cdots & p_{nn}
\end{pmatrix}
$

wobei  $ \sum_{i=1}^{n} p_{ij} = 1 \quad \text{für alle } j \in \{1, 2, \ldots, n\}
$,

und  $
p_{ij} \geq 0 \quad \text{für alle } i, j \in \{1, 2, \ldots, n\}
$

### Aufgabe 1.2: Status Vektor und Multiplikation mit Markov Matrix[3][5]

Wenn eine Markov-Matrix $P$ mit einem Statusvektor $s$ multipliziert wird, ergibt sich ein neuer Statusvektor $s' $. Die allgemeine Formel sieht folgendermaßen aus:

$
P = \begin{pmatrix}
p_{11} & p_{12} & \cdots & p_{1n} \\
p_{21} & p_{22} & \cdots & p_{2n} \\
\vdots & \vdots & \ddots & \vdots \\
p_{n1} & p_{n2} & \cdots & p_{nn}
\end{pmatrix}
$

$
\mathbf{s} = \begin{pmatrix}
s_1 \\
s_2 \\
\vdots \\
s_n
\end{pmatrix}
$

Das Produkt $\mathbf{s'} = P \mathbf{s} $  ergibt den neuen Statusvektor $ \mathbf{s'} $:

$
\mathbf{s'} = \begin{pmatrix}
s_1' \\
s_2' \\
\vdots \\
s_n'
\end{pmatrix}
$

wobei

$
\mathbf{s'} = P \mathbf{s} = \begin{pmatrix}
p_{11} & p_{12} & \cdots & p_{1n} \\
p_{21} & p_{22} & \cdots & p_{2n} \\
\vdots & \vdots & \ddots & \vdots \\
p_{n1} & p_{n2} & \cdots & p_{nn}
\end{pmatrix}
\begin{pmatrix}
s_1 \\
s_2 \\
\vdots \\
s_n
\end{pmatrix}
$

### Aufgaben 1.3, 1.4, 1.5: Analog zu Aufgabe 1.2 [4][6][7]

# Hilfsklasse für Aufgaben 1 und 2

In [2]:
class Krankheit_verlauf_2115():
    #Konstruktor
    def __init__(self):
        
        # Aufgabe 1.1
        #Krankheitsverlauf-Matrix
        self.uebergangsgesetz = [
 #           #G   #I   #K   #SK     #T #Im            
            [0.9, 0.03, 0,   0,     0,  0],#G         
            [0.1, 0.11, 0,   0,     0,  0],#I
            [0,   0.8,  0.4, 0.43,  0,  0],#K
            [0,   0.06, 0.1, 0.48,  0,  0],#SK
            [0,   0,    0,   0.09,  1,  0],#T
            [0,   0,    0.5, 0,     0,  1] #Im
        ]
        
        #Status der Bevölkerung am Anfang
        self.n_population = 50000
        self.n_immun = 1500
        self.infiziert_status = 1  # Annahme: Beginn mit 1 infizierten Person (Patient 0)
        self.gesund_status = self.n_population - self.n_immun - self.infiziert_status 
        self.krank_status = 0
        self.schwerkrank_status = 0
        self.tot_status = 0
        
        
        #Status-Vektor
        self.start_status =[
            self.gesund_status, 
            self.infiziert_status, 
            self.krank_status, 
            self.schwerkrank_status, 
            self.tot_status,
            self.n_immun
        ]
    
    
    #Simulation der Epidemie mit Status Vektor
    #Status Vektor wird zum ausgewälten Zeitpunkt zurückgegeben
    def simulation(self, wochen):
        M = Matrix(self.uebergangsgesetz)
        v = vector(self.start_status)
        for i in range(wochen):
            v = M * v
        return v
    
    
    #Simulation der Epidemie mit Status Vektor
    #Status Vektor wird zum ausgewälten Zeitpunkt ausgegeben
    def pretty_simulation(self, wochen):
        M = Matrix(self.uebergangsgesetz)
        v = vector(self.start_status)
        for i in range(wochen):
            v = M * v
        print(f"gesund: {round(v[0])}, infiziert: {round(v[1])}, krank: {round(v[2])}, schwer krank: {round(v[3])}, tot: {round(v[4])}, immun: {round(v[5])}")
    
    
    #Aufgabe 2.1
    def epidemie_ueberstanden(self):
        M = Matrix(self.uebergangsgesetz)
        v = vector(self.start_status)
        woche = 0
        i=0
        while v[2] + v[3] >= 0.001 * self.n_population or i<=3:  # Summe der Kranken und Schwerkranken
            v = M * v
            woche += 1
            if i <=3:
                i+=1
        return woche
    
    
    #Aufgabe 2.2
    def epidemie_tode(self):
        M = Matrix(self.uebergangsgesetz)
        v = vector(self.start_status)
        woche=0
        while v[0]+v[1]+v[2]+v[3] > 0.5: #Simulation der Epidemie bis es keine krankem Individuen mehr gibt
            v = M * v                    #Aufrunden auf 0.5
            woche+=1
        return round(v[4]),int(woche)
    
    
    #Aufgabe 2.3    
    def kosten_kranker(self):
        M = Matrix(self.uebergangsgesetz)
        v = vector(self.start_status)
        ende_epideemie = self.epidemie_tode()[1]
        kosten=0
        kranker = 20 
        schwer_kranker = 150
        for i in range(ende_epideemie):
            kosten += v[2]*kranker + v[3]*schwer_kranker
            v= M * v
        return round(kosten,2)

#Epidemie erstellen mit Konstruktor    
epidemie = Krankheit_verlauf_2115()             

## Einsetzen in die Problemstellung

### 1.2 Krankheitsverteilung nach vier Wochen


In [3]:
print("Nach 4 Wochen sieht die Bevölkerung wie folgend aus:")
epidemie.pretty_simulation(4)

Nach 4 Wochen sieht die Bevölkerung wie folgend aus:
gesund: 32205, infiziert: 4056, krank: 6177, schwer krank: 1223, tot: 100, immun: 6239


### 1.3 Tode nach einem Jahr


In [4]:
tode_nach_Jahr = epidemie.simulation(52)[4]
print(f"Nach einem Jahr sterben {int(tode_nach_Jahr)} Menschen.")

Nach einem Jahr sterben 2169 Menschen.



### 1.4 Diagramm für Epidemieverlauf in den ersten zehn Wochen


In [5]:
import matplotlib.pyplot as plt

# Initialisieren der Listen, um die Daten für jede Woche zu sammeln
gesund = []
infiziert = []
krank = []
schwer_krank = []
tot = []
immun = []

# Daten für jede Woche sammeln
for woche in range(11):  # 0 bis 10 Wochen
    ergebnisse = epidemie.simulation(woche)
    gesund.append((woche, round(ergebnisse[0]))) # round(), um Rundungsfehler zu vermeiden
    infiziert.append((woche, round(ergebnisse[1])))
    krank.append((woche, round(ergebnisse[2])))
    schwer_krank.append((woche, round(ergebnisse[3])))
    tot.append((woche, round(ergebnisse[4])))
    immun.append((woche, round(ergebnisse[5])))

def plot_epidemie_data(wochen=1):
    # Daten für die gewünschte Woche sammeln
    woche_data = {
        "Gesund": gesund[:wochen+1],
        "Infiziert": infiziert[:wochen+1],
        "Krank": krank[:wochen+1],
        "Schwer krank": schwer_krank[:wochen+1],
        "Tot": tot[:wochen+1],
        "Immun": immun[:wochen+1]
    }

    # Aktuelle Woche und Anzahl der betroffenen Personen drucken
    aktuelle_werte = {
        "Gesund": gesund[wochen][1],
        "Infiziert": infiziert[wochen][1],
        "Krank": krank[wochen][1],
        "Schwer krank": schwer_krank[wochen][1],
        "Tot": tot[wochen][1],
        "Immun": immun[wochen][1]
    }

    print(f"Woche {wochen}:")
    for zustand, anzahl in aktuelle_werte.items():
        print(f"{zustand}: {anzahl}")

    # Diagramm zeichnen
    plt.figure(figsize=(10, 6))

    # Zeichnen jeder Kategorie
    plt.plot(*zip(*woche_data["Gesund"]), color="lightgreen", label="Gesund", marker='o', linestyle="--")
    plt.plot(*zip(*woche_data["Infiziert"]), color="yellow", label="Infiziert", marker='s', linestyle="-.")
    plt.plot(*zip(*woche_data["Krank"]), color="orange", label="Krank", marker='^', linestyle=":")
    plt.plot(*zip(*woche_data["Schwer krank"]), color="red", label="Schwer krank", marker='v', linestyle="-")
    plt.plot(*zip(*woche_data["Tot"]), color="black", label="Tot", marker='x', linestyle="--")
    plt.plot(*zip(*woche_data["Immun"]), color="darkgreen", label="Immun", marker='p', linestyle="-")

    plt.xlabel('Zeit in Wochen')
    plt.ylabel('Einwohnerzahl')
    plt.title('Das Beteigeuze-Fieber')
    plt.legend(loc='upper right')
    plt.grid(True)
    plt.xlim(0, 10)  # Bereich der x-Achse auf 10 Wochen beschränken
    plt.show()

# Interaktive Funktion mit Slider zur Auswahl der Woche
@interact
def interactive_plot(wochen=slider(0, 10, 1, default=1, label="Wochen")):
    plot_epidemie_data(int(wochen))



Interactive function <function interactive_plot at 0x117781c60> with 1 widget
  wochen: TransformIntSlider(value=1, description='Wochen', max=10)


### 1.5 Animation für die Bevölkerungsverteilung auf die Stadien für das erste Jahr

In [6]:
def animate_population(wochen=1):
    status = epidemie.simulation(wochen)
    categories = ["Gesund", "Infiziert", "Krank", "Schwer kr.", "Tot", "Immun"]
    colors = ['lightgreen', 'yellow', 'orange', 'red', 'black', 'darkgreen']  # Farben definieren

    # Balkendiagramm
    fig, ax = plt.subplots()
    bars = ax.bar(categories, status, width=0.8, color=colors)

    # Textlabels über jeder Bar für die Anzahl betroffener Personen
    for i, rect in enumerate(bars):
        height = rect.get_height()
        ax.text(rect.get_x() + rect.get_width() / 2., 1.05 * height, f'{round(status[i])}',
                ha='center', va='bottom', fontsize=10, color='black')

    # Entfernen von x-Achsenticks und setzen der Kategorienamen als Labels
    ax.set_xticks(range(len(categories)))  
    ax.set_xticklabels(categories)

    # Y-Achse beschriften
    ax.set_ylabel("Anzahl betroffener Personen")

    # y-Achsenbereich festlegen und Rahmen anzeigen
    ax.set_ylim(0, epidemie.n_population)
    ax.set_frame_on(True)

    plt.show()

# Interaktionsfunktion mit einem Slider, der es erlaubt, die Woche einzustellen
@interact(wochen=slider(0, 52, 1, 0))
def interactive_plot(wochen):
    animate_population(wochen)

Interactive function <function interactive_plot at 0x117880720> with 1 widget
  wochen: TransformIntSlider(value=0, description='wochen', max=52)

# Lösungsansätze
# Aufgabe 2 – Bestimmen Sie das Ende der Epidemie

### 2.1 Überstehen der Epidemie (Anzahl Erkrankte < 0,1 %)
- Wir simulieren den Verlauf der Epidemie mithilfe einer Hilfsfunktion, bis der Anteil der kranken und schwer kranken Individuen unter 0,1 % liegt.
- Das Ergebnis ist die Woche, in der dieser Zustand erreicht wird.

**siehe Hilfsklasse**

### 2.2 Todesfälle nach biologischem Ende der Epidemie
**Annahme**: Die Epidemie gilt als beendet, wenn alle Individuen entweder tot oder immun sind.[9]
- Wir simulieren den Verlauf der Epidemie mithilfe einer Hilfsfunktion, bis nur noch tote oder immune Individuen verbleiben.
- Das Ergebnis ist die Woche, in der wir das biologische Ende der Epidemie erreichen.

**siehe Hilfsklasse**

### 2.3 Kosten für Kranke im Verlauf der Epidemie
- Wir simulieren den Verlauf der Epidemie bis zum biologischen Ende. Dabei erfassen wir wöchentlich die Anzahl der kranken und schwer kranken Individuen und multiplizieren diese mit den entsprechenden Kosten.
- Das Ergebnis sind die Gesamtkosten am Ende der Epidemie.


**siehe Hilfsklasse**

## Was verwenden wir, um das Problem zu lösen?
- Analog zu Aufgabe 1 verwenden wir die Multiplikation der Markov-Matrix $P$ mit dem Statusvektor $s$, um den Verlauf der Epidemie zu simulieren.[1][3][5]


$
\mathbf{s'} = P \mathbf{s} = \begin{pmatrix}
p_{11} & p_{12} & \cdots & p_{1n} \\
p_{21} & p_{22} & \cdots & p_{2n} \\
\vdots & \vdots & \ddots & \vdots \\
p_{n1} & p_{n2} & \cdots & p_{nn}
\end{pmatrix}
\begin{pmatrix}
s_1 \\
s_2 \\
\vdots \\
s_n
\end{pmatrix}
$


## Einsetzen in die Aufgabenstellung
### 2.1 Ende der Epidemie

In [7]:
print(f"Die Epidemie ist nach {epidemie.epidemie_ueberstanden()} Wochen überstanden.")

Die Epidemie ist nach 58 Wochen überstanden.


### 2.2 Gesamtzahl der Todesfälle

In [8]:
print(f"Es sterben insgesamt {epidemie.epidemie_tode()[0]} Individuen nach {epidemie.epidemie_tode()[1]} Wochen.")

Es sterben insgesamt 2189 Individuen nach 118 Wochen.


### 2.3 Kosten der Epidemie

In [9]:
print(f"Die Gesamtkosten für Erkranke nach der Epidemie betragen {epidemie.kosten_kranker()} Space Coins")

Die Gesamtkosten für Erkranke nach der Epidemie betragen 5500270.78 Space Coins


# Lösungsansätze

# Aufgabe 3 – Einfluss von Impfung

## Annahmen[2][8]

- Gesunde, infizierte, kranke und schwer kranke Individuen werden geimpft.
- Immune und verstorbene Individuen werden nicht geimpft.
- Durch die Impfung verringert sich die Wahrscheinlichkeit um 50%, in den infizierten, kranken, schwer kranken oder toten Status überzugehen.
- Die Impfung beginnt ab der fünften Woche.
- Die Impfrate beträgt:
    1. 5% bei gesunden Individuen
    2. 3% bei infizierten Individuen
    3. 2% bei kranken Individuen
    4. 1% bei schwer kranken Individuen

## Allgemein

- Es wird eine Subklasse erstellt, die den Verlauf der Epidemie unter Berücksichtigung der Impfung simuliert.
- Folgende Stadien werden der Übergangsmatrix hinzugefügt:
    1. gesund und geimpft
    2. infiziert und geimpft
    3. krank und geimpft
    4. schwer krank und geimpft
- Die oben genannten Stadien werden auch dem Statusvektor hinzugefügt.
- Es werden Hilfsfunktionen definiert, die den Verlauf der Epidemie mit Impfung simulieren.
- Diese Hilfsfunktionen werden verwendet, um die verschiedenen Szenarien zu vergleichen.

## Ziel der Auswertung

- Ziel ist es, die verschiedenen Szenarien mit und ohne Impfung anhand der folgenden Kriterien zu bewerten:
    1. Offizielle Epidemiedauer (<0,1%)
    2. Biologische Epidemiedauer
    3. Anzahl der Todesfälle
    4. Gesamtkosten
  

# Hilfsklasse für die Simulation der Epidemie mit Impfung


In [10]:
class Krankheit_verlauf_mit_Impfung(Krankheit_verlauf_2115):
    def __init__(self):
        super().__init__()
        
        # Modifikation der Übergangsmatrix für die Auswirkungen der Impfung
        # ' -> Impfung
        self.uebergangsgesetz = [
#            G    I     K    SK     T  Im   G'      I'     K'    SK' 
            [0.9, 0.03, 0,   0,     0, 0,   0,      0,     0,    0      ], #G
            [0.1, 0.11, 0,   0,     0, 0,   0,      0,     0,    0      ], #I
            [0,   0.8,  0.4, 0.43,  0, 0,   0,      0,     0,    0      ], #K
            [0,   0.06, 0.1, 0.48,  0, 0,   0,      0,     0,    0      ], #SK
            [0,   0,    0,   0.09,  1, 0,   0,      0,     0,    0.045  ], #T
            [0,   0,    0.5, 0,     0, 1,   0,      0,     0.75, 0      ], #Im
            [0,   0,    0,   0,     0, 0,   0.95,   0.515, 0,    0      ], #G'
            [0,   0,    0,   0,     0, 0,   0.05,   0.055, 0,    0      ], #I'
            [0,   0,    0,   0,     0, 0,   0,      0.4,   0.2,  0.715  ], #K'
            [0,   0,    0,   0,     0, 0,   0,      0.03,  0.05, 0.24   ], #SK'
        ]
        
        #Status mit Berücksichtigung von Impfung
        self.gesund_status_geimpft = 0
        self.infizier_status_geimpft = 0
        self.krank_status_geimpft = 0
        self.schwerkrank_status_geimpft = 0
        
        # Erweiterter Status-Vektor mit Impfung
        self.start_status.extend([
            self.gesund_status_geimpft,
            self.infizier_status_geimpft,
            self.krank_status_geimpft,
            self.schwerkrank_status_geimpft
        ])
    
    # Impfvorgang
    def impfen(self, v):
        impfung_gesunde = v[0] * 0.05
        impfung_infizierte = v[1] * 0.03
        impfung_kranke = v[2] * 0.02
        impfung_schwerkranke = v[3] * 0.01        
        v[0] -= impfung_gesunde
        v[1] -= impfung_infizierte
        v[2] -= impfung_kranke
        v[3] -= impfung_schwerkranke 
        v[6] += impfung_gesunde
        v[7] += impfung_infizierte
        v[8] += impfung_kranke
        v[9] += impfung_schwerkranke
        return vector(v)
                
    def simulation(self, wochen):
        M = Matrix(self.uebergangsgesetz)
        v = vector(self.start_status)
        for i in range(wochen):
            if i >= 4:
                v = self.impfen(v)               
            v = M * v
        return v
    
    def pretty_simulation(self, wochen):
        M = Matrix(self.uebergangsgesetz)
        v = vector(self.start_status)
        for i in range(wochen):
            if i >= 4:
                v = self.impfen(v)                
            v = M * v
        print(f"gesund: {round(v[0])}, infiziert: {round(v[1])}, krank: {round(v[2])}, schwer krank: {round(v[3])}, tot: {round(v[4])}, immun: {round(v[5])}, gesund geimpft: {round(v[6])}, infiziert geimpft: {round(v[7])}, krank geimpft: {round(v[8])}, schwer krank geimpft: {round(v[9])}")
        
    def epidemie_ueberstanden_impfung(self):
        M = Matrix(self.uebergangsgesetz)
        v = vector(self.start_status)
        woche = 0
        i=0
        while v[2] + v[3] + v[8] + v[9] >= 0.001 * self.n_population or i<=3:  # Summe der Kranken und Schwerkranken mit Berücksichtigung der Impfung
            if woche >= 4:
                v = self.impfen(v)
            v = M * v
            woche += 1
            if i <=3:
                i+=1
        return woche
    
    
    # Biologisches Ende
    def epidemie_tode_impfung(self):
        M = Matrix(self.uebergangsgesetz)
        v = vector(self.start_status)
        woche=0
        while v[0]+v[1]+v[2]+v[3]+v[6]+v[7]+v[8]+v[9] > 1:
            if woche >= 4:
                v = self.impfen(v)                
            v = M * v
            woche+=1
        return round(v[4]),woche

    # Kostenrechnung
    def kosten_kranker_impfung(self):
        kranker_kosten_pro_woche = 20
        schwerkranker_kosten_pro_woche = 150
        impfung_kosten = 15

        M = Matrix(self.uebergangsgesetz)
        v = vector(self.start_status)
        gesamtkosten = 0
        impfung_gesamtkosten = 0

        for woche in range(self.epidemie_tode_impfung()[1]):  # Berechnung der Impfung-Kosten 
            if woche >= 4:
                
                impfung_gesunde = v[0] * 0.05
                impfung_infizierte = v[1] * 0.03
                impfung_kranke = v[2] * 0.02
                impfung_schwerkranke = v[3] * 0.01
                impfung_gesamtkosten += (impfung_gesunde + impfung_infizierte + impfung_kranke + impfung_schwerkranke) * impfung_kosten
                
                v = self.impfen(v)
            v = M * v
            # Kranken-Kosten pro Woche
            woche_kosten = (v[2] * kranker_kosten_pro_woche + 
                            v[3] * schwerkranker_kosten_pro_woche + 
                            v[8] * kranker_kosten_pro_woche + 
                            v[9] * schwerkranker_kosten_pro_woche)
            gesamtkosten += woche_kosten

        gesamtkosten += impfung_gesamtkosten
        
        return round(gesamtkosten, 2), round(impfung_gesamtkosten, 2)
    
        
epidemie_impfung = Krankheit_verlauf_mit_Impfung()

## Auswertung der Epidemie mit Impfung
### Simulation


In [11]:
# Funktion zur Animation der Population
def animate_population_impfung(wochen=1):
    status = epidemie_impfung.simulation(wochen)
    kategorien = ["Gesund", "Infiziert", "Krank", "Schwer kr.", "Tot", "Immun", "Gesund geimpft", "Infiziert geimpft", "Krank geimpft", "Schwer kr. geimpft"]
    farben = ['lightgreen', 'yellow', 'orange', 'red', 'black', 'darkgreen', 'lightgreen', 'yellow', 'orange', 'red']  # Farben definieren

    # Balkendiagramm
    fig, ax = plt.subplots(figsize=(12, 6))
    bars = ax.bar(kategorien, status, width=0.5, color=farben)

    # Textlabels über jeder Bar für die Anzahl betroffener Personen
    for i, rect in enumerate(bars):
        height = rect.get_height()
        ax.text(rect.get_x() + rect.get_width() / 2., 1.05 * height, f'{round(status[i])}',
                ha='center', va='bottom', fontsize=10, color='black')

    # Kategorienbeschriftungen rotieren und abstand vergrößern
    ax.set_xticks(range(len(kategorien)))  
    ax.set_xticklabels(kategorien, rotation=45, ha='right')

    # Y-Achse beschriften
    ax.set_ylabel("Anzahl betroffener Personen")

    # y-Achsenbereich festlegen und Rahmen anzeigen
    ax.set_ylim(0, epidemie_impfung.n_population)
    ax.set_frame_on(True)

    plt.tight_layout()
    plt.show()

# Interaktionsfunktion mit einem Slider, der es erlaubt, die Woche einzustellen
@interact
def interactive_plot(wochen=slider(0, 433, 1, default=0)):
    animate_population_impfung(wochen)



Interactive function <function interactive_plot at 0x1178d8400> with 1 widget
  wochen: TransformIntSlider(value=0, description='wochen', max=433)

## Offizielles Ende (<0.1%)

In [12]:
print(f"Die Epidemie ist nach {epidemie_impfung.epidemie_ueberstanden_impfung()} Wochen überstanden")

Die Epidemie ist nach 106 Wochen überstanden


## Biologisches Ende

In [13]:
print(f"Es sterben insgesamt {epidemie_impfung.epidemie_tode_impfung()[0]} Individuen nach {epidemie_impfung.epidemie_tode_impfung()[1]} Wochen.")

Es sterben insgesamt 1677 Individuen nach 433 Wochen.


## Gesamtkosten

In [14]:
print(f"Die Gesamtkosten der Epidemie mit Einsatz einer Impfung betragen {round(epidemie_impfung.kosten_kranker_impfung()[0],2)} Space Coins.")
print(f"Die Impfung alleine kostet für die Epidemie {round(epidemie_impfung.kosten_kranker_impfung()[1],2)} Space Coins. " )

Die Gesamtkosten der Epidemie mit Einsatz einer Impfung betragen 4881438.91 Space Coins.
Die Impfung alleine kostet für die Epidemie 203835.53 Space Coins. 


## Auswertungstabelle "Ohne Impfung vs Mit Impfung"

|                           | Ohne Impfung  |  Mit Impfung           |
|-----------------------|---------------|------------------------|
| offizielle Epidemie Dauer (<0.1%) | 58 Wochen| 106 Wochen                    |
| biologische Epidemie Dauer | 118 Wochen            | 433 Wochen                   |
| Anzahl Tode               | 2189            | 1677|
| Gesamtkosten              | 5500270.78          | 4881438.91                      | 


## Auswertung

### Epidemie Dauer:

- Die offizielle Epidemie Dauer, definiert als die Zeit, in der die Anzahl der kranken und schwer kranken Individuen über 0,1% der Population liegt, verlängert sich durch die Impfung von 58 Wochen auf 106 Wochen. Durch die Impfung erkranken Menschen seltener, was dazu führt, dass weniger Menschen durch eine Erkrankung immun werden.
- Die biologische Epidemie Dauer, die Zeit bis alle Individuen entweder tot oder immun sind, verlängert sich drastisch von 118 Wochen ohne Impfung auf 433 Wochen mit Impfung.

### Anzahl Tode:

- Die Anzahl der Todesfälle sinkt deutlich durch den Einsatz der Impfung. Ohne Impfung sterben 2189 Menschen, während mit Impfung die Anzahl der Todesfälle auf 1677 reduziert wird. Dies zeigt, dass die Impfung effektiv dazu beiträgt, Leben zu retten, auch wenn die Epidemie länger dauert.

### Gesamtkosten:

- Trotz der verlängerten Dauer der Epidemie sind die Gesamtkosten mit Impfung geringer. Ohne Impfung betragen die Kosten 5.500.270,78 Coins, während sie mit Impfung auf 4881438,91 Coins sinken. Die Impfung führt zu einer Reduktion der Behandlungskosten für kranke und schwer kranke Individuen, was die zusätzlichen Impfkosten mehr als ausgleicht.

## Fazit:

- Die Impfung verlängert zwar die Dauer der Epidemie erheblich, führt jedoch zu einer signifikanten Reduktion der Todesfälle und der Gesamtkosten. Diese Ergebnisse unterstreichen die Bedeutung der Impfung als wirksame Maßnahme im Kampf gegen Epidemien. Die verlängerte Epidemiedauer ist ein akzeptabler Kompromiss angesichts der erheblichen Vorteile in Bezug auf gerettete Leben und gesenkte Kosten.

# Unit Testing
### Test 1: Population Anzahl soll immer 50.000 betragen
- **Ziel**: Rundungsfehler vermeiden
- **Zum Testen**: Wochenanzahl anpassen


In [15]:
epidemie_test = Krankheit_verlauf_2115()
summe = 0
woche = 52
for i in range(6):
    summe += epidemie_test.simulation(woche)[i]
    print(f"Iteration {i}")
    print(f"Eintrag {i} Status Vektor: {round(epidemie_test.simulation(woche)[i],2)}" )
    print(f"Aufsummiert: {round(summe,2)}")
    print("-------------")
    
print(f"Summe: {summe}")
print(f"Test: {round(summe,5) == 50000}")
    

Iteration 0
Eintrag 0 Status Vektor: 250.56
Aufsummiert: 250.56
-------------
Iteration 1
Eintrag 1 Status Vektor: 31.57
Aufsummiert: 282.13
-------------
Iteration 2
Eintrag 2 Status Vektor: 67.55
Aufsummiert: 349.67
-------------
Iteration 3
Eintrag 3 Status Vektor: 20.41
Aufsummiert: 370.08
-------------
Iteration 4
Eintrag 4 Status Vektor: 2169.64
Aufsummiert: 2539.72
-------------
Iteration 5
Eintrag 5 Status Vektor: 47460.28
Aufsummiert: 50000.0
-------------
Summe: 50000.0000000000
Test: True


In [16]:
epidemie_impfung_test = Krankheit_verlauf_mit_Impfung()
summe = 0
woche = 52
for i in range(10):
    summe += epidemie_impfung_test.simulation(woche)[i]
    print(f"Iteration {i}")
    print(f"Eintrag {i} Status Vektor: {round(epidemie_impfung_test.simulation(woche)[i],2)}" )
    print(f"Aufsummiert: {round(summe,2)}")
    print("-------------")
    
print(f"Summe: {summe}")
print(f"Test: {round(summe,5) == 50000}")

Iteration 0
Eintrag 0 Status Vektor: 21.46
Aufsummiert: 21.46
-------------
Iteration 1
Eintrag 1 Status Vektor: 2.71
Aufsummiert: 24.18
-------------
Iteration 2
Eintrag 2 Status Vektor: 6.37
Aufsummiert: 30.55
-------------
Iteration 3
Eintrag 3 Status Vektor: 2.04
Aufsummiert: 32.59
-------------
Iteration 4
Eintrag 4 Status Vektor: 1635.21
Aufsummiert: 1667.79
-------------
Iteration 5
Eintrag 5 Status Vektor: 43357.46
Aufsummiert: 45025.25
-------------
Iteration 6
Eintrag 6 Status Vektor: 4562.19
Aufsummiert: 49587.44
-------------
Iteration 7
Eintrag 7 Status Vektor: 247.17
Aufsummiert: 49834.61
-------------
Iteration 8
Eintrag 8 Status Vektor: 145.47
Aufsummiert: 49980.08
-------------
Iteration 9
Eintrag 9 Status Vektor: 19.92
Aufsummiert: 50000.0
-------------
Summe: 50000.0000000000
Test: True


# Anhang

In [17]:
from pybtex.database import parse_file
import re

from pybtex.database import parse_file
import re

bibfile = "Projekt2.bib"

def clean_latex_formatting(text):
    text = re.sub(r"(\{|\})", "", text)
    return text

def format_ieee(entry, entry_number):
    # Title bereinigen
    title = clean_latex_formatting(entry.fields.get('title', '[Title not found]'))
    
    # Autorenliste erstellen, falls vorhanden
    authors = entry.persons.get('author', None)
    if authors:
        authors = ' and '.join(' '.join(p.first_names + p.last_names) for p in authors)
    else:
        authors = title  # Use title if author is not available
    
    year = entry.fields.get('year', '[Year not found]')
    url = entry.fields.get('url', '')

    # Formatieren basierend auf dem Eintragstyp
    if 'booktitle' in entry.fields:  # for conference papers
        booktitle = clean_latex_formatting(entry.fields.get('booktitle'))
        citation = f"[{entry_number}] {authors}, \"{title},\" in {booktitle}, {year}."
    if 'wikipedia.org' in url:
        # Formatting Wikipedia-specific reference
        citation = f"[{entry_number}] \"{title},\" Wikipedia, {year}. [Online]. Available: {url}."
    elif 'journal' in entry.fields:  # for journal articles
        journal = clean_latex_formatting(entry.fields.get('journal'))
        volume = entry.fields.get('volume', '[Volume not found]')
        number = entry.fields.get('number', '[Issue not found]')
        pages = entry.fields.get('pages', '[Pages not found]')
        citation = f"[{entry_number}] {authors}, \"{title},\" {journal}, vol. {volume}, no. {number}, pp. {pages}, {year}."
    elif entry.type == 'book':  # for books
        publisher = clean_latex_formatting(entry.fields.get('publisher', '[Publisher not found]'))
        citation = f"[{entry_number}] {authors}, {title}. {publisher}, {year}."
    else:  # for other types
        citation = f"[{entry_number}] {authors}, {title}, {year}."

    return citation

# BibTeX-Datei lesen
bib_data = parse_file(bibfile)

# Alle Einträge im IEEE-Stil konvertieren
ieee_references = [format_ieee(entry, i+1) for i, (key, entry) in enumerate(bib_data.entries.items())]

# Alle IEEE-Stil Zitate ausgeben
print("\033[1mQuellen\033[0m ")
print('')
for ref in ieee_references:
    print(ref)
    print('')

Quellen 

[1] "Übergangsmatrix," Wikipedia, [Year not found]. [Online]. Available: https://de.wikipedia.org/w/index.php?title=%C3%9Cbergangsmatrix&oldid=243066595.

[2] Bundesministerium Gesundheit, Das offizielle Dashboard zur Impfkampagne der Bundesrepublik Deutschland, [Year not found].

[3] Matrix mal Vektor • Berechnung und Beispiele, Matrix mal Vektor • Berechnung und Beispiele, [Year not found].

[4] 2D plotting - 2D Graphics, 2D plotting - 2D Graphics, [Year not found].

[5] Thorsten Imkamp, Einstieg in stochastische Prozesse. Springer Spektrum, [Year not found].

[6] Bar charts - 2D Graphics, Bar charts - 2D Graphics, [Year not found].

[7] Animated plots - 2D Graphics, Animated plots - 2D Graphics, [Year not found].

[8] Zentralinstitut Bundesrepublik Deutschland, Modellierung von Impfszenarien, [Year not found].

[9] Bayerischer Heike Westram, Pocken, Grippe, Cholera, HIV: Wann endet eine Pandemie?, [Year not found].

